### DISCUSSION TITLE
Connection reset by peer

### DISCUSSION THREAD

**Discussion by Bilge** [deployphp/deployer#2297] [General]
Deployment frequently fails, during any task at any time during the deployment pipeline, with a _connection reset by peer_ message. I have no issues having a PuTTY connection open to the same server for days without interruption but deployer cannot stay connected for more than a few seconds.

```
  mm_send_fd: sendmsg(2): Connection reset by peer
  mux_client_request_session: send fds failed
```

Eventually, after retrying 12 times and failing every single time at random points during the deploy, it got to the point where it refused to even start. 

>➤ Executing task deploy:prepare

It just gets stuck at the first task and emits no further output, not even a connection reset by peer. Deployer is now completely broken in that shell and I had to launch another shell to get it to respond again, but it still gets connection reset.

Windows 10 over IPv6 using v6.8.0.

If I had to guess, it might be something to do with the ssh-pageant agent.

```bash
# ssh-pageant allows use of the PuTTY authentication agent (Pageant)
SSH_PAGEANT="$(command -v ssh-pageant)"
if [ -x "$SSH_PAGEANT" ]; then
   eval $("$SSH_PAGEANT" -qra "${SSH_AUTH_SOCK:-${TEMP:-/tmp}/.ssh-pageant-$USERNAME}")
fi
unset SSH_PAGEANT
```

I use this script to connect SSH to Pageant so I do not have to specify my keys.

**Comment by antonmedv** [comment]
Try to omit pagent to check if it is the case.

**Reply by Bilge** [reply]
It isn't to do with Pageant. I removed `SSH_AUTH_SOCK`, and of course, I get *Permission denied (publickey).* This confirms I removed the link to Pageant successfully. Then I do `eval $(ssh_agent)` and `ssh-add` my key, to use the normal SSH agent instead. Immediately I get hit with another:

> mm_send_fd: sendmsg(2): Connection reset by peer
  mux_client_request_session: send fds failed

It seems to be linked to MINGW64/mintty/Git bash environment on Windows rather than the specific SSH authentication agent being used.

---

Confirmed working if I use WSL instead of Git bash. Not only does it work under WSL but it is also dramatically faster, like at least 2x faster. But I don't want to use WSL.

**Reply by antonmedv** [reply]
Why don’t you want to use WSL?

**Reply by Bilge** [reply]
WSL is not part of my workflow and has poor integration with Windows, including filesystem navigation and clipboard shortcuts. I do everything in Git bash: it has better integration with Windows and my workflow toolkit plus it looks better.

| Git Bash| WSL |
|-|-|
| ![image](https://user-images.githubusercontent.com/470626/100390462-59e5f580-3028-11eb-9452-43727d4c6e53.png) | ![image](https://user-images.githubusercontent.com/470626/100390567-c660f480-3028-11eb-8c66-1a69fd74b488.png)

I definitely don't want to have to be switching into WSL just to deploy.
